### ISSUE TITLE
L4显卡 TRT 10.16 环境下，engine模型使用tensorRT推理的精度和onnx及pt模型对不上

### ISSUE THREAD

**Issue by kilakila-heart** [Intellindust-AI-Lab/DEIMv2#152]
我们使用了pt模型、onnx模型和engine模型，发现onnx精度可以对齐pt模型，但是engine模型的精度和他们相差很大，请问大佬你们是如何转换的，有什么技巧吗？
PT和onnx对齐度很高，这是compare_raw_outputs.py的结果： [PT vs ONNX]
    logits cosine:      0.99999753
    平均 Box IOU:        0.9998
    IOU ≥ 0.9 占比:      100.0%
    Top-5 预测重叠率:    100%
    ✅ 精度完好，可直接上线，但是pt和trt的对齐度很低，这是compare_pt_trt.py的结果：================================================================================

### 环境

- **TRT 版本**: 10.16.1.11
- **PyTorch 版本**: 2.x (export with `torch.onnx.export`)
- **ONNX Runtime**: 1.x (CPU EP)
- **CUDA**: 12.x, cuDNN 8.x
- **GPU**: （你的 GPU 型号）

### 模型

- **架构**: DEIMv2 + DINOv3-X backbone (Shein 配置)
- **Backbone**: DINOv3 ViT-X (~24 transformer blocks, RoPE + self-attention + LayerNorm)
- **Encoder**: HybridEncoder (FPN + PAN + TransformerEncoder)
- **Decoder**: DEIMv2 Decoder
- **输入**: [B, 3, 640, 640]
- **输出**: pred_logits [B, 300, 23], pred_boxes [B, 300, 4]
- **ONNX 导出**: opset 17, `do_constant_folding=True`, 含 backbone+encoder+decoder（不含 PostProcessor）

---

### 实验 1：端到端精度对比（PT vs ONNX RT vs TRT）

对比 PyTorch → ONNX Runtime 和 ONNX Runtime → TRT 的端到端输出（pred_logits + pred_boxes）。

| 对比 | Box Mean IOU | IOU > 0.9 | Top-5 重叠率 |
|------|-------------|-----------|-------------|
| PT vs ONNX RT | ~1.0 | ~100% | ~100% |
| PT vs TRT (FP32 默认) | 0.8495 | 74.4% | 42% |

**结论**: PT→ONNX 精度近乎完美，PT→TRT 精度损失严重。

---

### 实验 2：TRT Builder 配置调优（端到端）

在**端到端模型**上尝试不同 TRT builder 配置，对比 PT vs TRT：

| 配置 | Box Mean IOU | IOU > 0.9 | Top-5 重叠率 |
|------|-------------|-----------|-------------|
| 默认 FP32 (baseline) | 0.8495 | 74.4% | 42% |
| + `restrict_tactics` (仅 cuBLAS + CUBLAS_LT) | 0.8745 | 80.8% | 47% |
| + `prefer_precision` (PREFER + OBEY precision constraints) | 0.8986 | 83.6% | 50% |
| + `builder_opt_level=0` (最低优化) | 0.8975 | 80.0% | 53% |

**结论**: 所有 TRT 配置都有显著精度损失，最优组合也仅到 IOU 0.90，Top-5 仅 50%。

---

### 实验 3：隔离 Backbone vs Encoder+Decoder

为定位精度损失来源，单独导出 backbone（DINOv3STAs）为 ONNX，分别用 ORT 和 TRT 运行，对比 backbone 输出的多尺度特征图（c2 [B,256,80,80], c3 [B,256,40,40], c4 [B,256,20,20]）。

#### 3a：PT vs ONNX RT（backbone 基线）

| 层级 | cosine_sim | max_abs_diff | mean_abs_diff |
|------|-----------|-------------|--------------|
| c2   | ~1.000000 | 5.6~7.1e-03 | ~5.3e-04 |
| c3   | ~1.000000 | 2.6~6.1e-03 | ~3.0e-04 |
| c4   | ~1.000000 | 2.4~2.6e-03 | ~2.5e-04 |

**PT→ONNX 精度近乎完美**，证明 ONNX 图本身无问题。

#### 3b：ONNX RT vs TRT（backbone，默认配置）

5 张图片，ONNX vs TRT 的 c3 层级（误差最大层）：

| 图片 | max_abs_diff | mean_abs_diff | 说明 |
|------|-------------|--------------|------|
| 0    | 0.115       | 0.0035       | 正常 |
| 1    | 0.858       | 0.0039       | 严重 |
| 2    | 0.201       | 0.0026       | 中等 |
| 3    | 0.184       | 0.0029       | 中等 |
| 4    | **1.644**   | 0.0039       | **极严重** |

**问题确认在 backbone 内部**。误差高度依赖输入图片：图片 4 的 c3 误差是图片 0 的 14 倍。所有图片呈现"mean 小、max 大"的模式 —— 绝大多数值接近，少数值点偏差极大。

---

### 实验 4：TRT Tactic Source 隔离（backbone）

在 backbone-only ONNX 上试不同 tactic source 配置：

| 配置 | ONNX vs TRT c3 max_abs_diff（5 图） | vs 默认 |
|------|-------------------------------------|---------|
| 默认 (全 tactic) | 0.115 / 0.858 / 0.201 / 0.184 / 1.644 | 基线 |
| `--cublas-only` (仅 CUBLAS) | 0.129 / 0.851 / 0.207 / 0.174 / 1.636 | **无改善** |

禁用 CUBLAS_LT（fused attention kernel 来源）**无效果**，说明问题不在 fused attention kernel。

---

### 实验 5：更换 ONNX Opset 版本（backbone）

用 opset 14 重新导出 backbone ONNX，构建 TRT engine 后对比：

| 配置 | ONNX vs TRT c3 max_abs_diff（5 图） | vs opset 17 |
|------|-------------------------------------|-------------|
| opset 17 (默认) | 0.115 / 0.858 / 0.201 / 0.184 / 1.644 | 基线 |
| opset 14 | 0.115 / 0.858 / 0.201 / 0.184 / 1.644 | **完全相同** |

opset 14 和 opset 17 的结果**逐位一致**，TRT 对不同 opset 的图收敛到相同的内部表示。

---

### 总结

1. **PT→ONNX (ORT CPU EP) 精度近乎完美**，ONNX 图本身正确。
2. **TRT 的精度损失来自 DINOv3 backbone**，与 encoder/decoder 无关。
3. **所有 TRT 配置均无效**：restrict_tactics、prefer_precision、builder_opt_level=0、cublas_only、opset 14 —— 没有任何一项能显著改善 backbone 的精度。
4. **误差与输入高度相关**：某些图片的 c3 max_abs_diff 达 1.64，是其他图片的 14 倍；mean_abs_diff 却始终很小（~0.004），说明是**少数特征值点的数值不稳定**，而非整体偏移。

---

**Comment by yuhyun00** [comment]
Hi, thanks for the detailed report!

This might be related to TF32. On Ampere or newer GPUs (L4 included, since it's Ada Lovelace), TensorRT enables TF32 by default even when you request FP32, which can cause the kind of small-but-scattered value errors you're seeing (small mean_abs_diff but large max_abs_diff). Could you check whether TF32 is disabled in your builder config? For example:

```python
config.clear_flag(trt.BuilderFlag.TF32)
```

You can see how this is handled in a similar reference implementation here: https://github.com/yuhyun00/DEIMv2-FastTracker-TensorRT/blob/main/export/onnx2trt.py

Might be worth trying and comparing PT vs TRT again to see if it closes the gap.

**Comment by Ahmedkhalilelzainy** [comment]
did you reach anything on this

**Comment by wad-glitch** [comment]
试一下将deformable_attention_core_func_v2函数（engine/deim/utils.py）中sampling_value_list列表进行物化并在后续加权时取消对batch和注意力头的合并
